index.d.ts

```diff
@@ -0,0 +1,30 @@
+import { Group } from 'ol/layer';
+import BaseLayer from 'ol/layer/Base';
+import Control, { Options as ControlOptions } from 'ol/control/Control';
+import PluggableMap from 'ol/PluggableMap';
+
+type GroupSelectStyle = 'none' | 'children' | 'group';
+
+export interface Options extends ControlOptions {
+    activationMode?: string;
+    collapseLabel?: string;
+    label?: string;
+    tipLabel?: string;
+    groupSelectStyle?: GroupSelectStyle;
+    reverse?: boolean
+}
+
+export default class LayerSwitcher extends Control {
+    constructor(options?: Options);
+    static isBaseGroup(lyr: BaseLayer): boolean;
+    static renderPanel(map: PluggableMap, panel: Element, options?: {groupSelectStyle?: GroupSelectStyle; reverse?: boolean;}): void;
+    static setGroupVisibility(map: PluggableMap): any[];
+    static setChildVisibility(map: PluggableMap): any[];
+    static forEachRecursive(lyr: Group, fn: (lry: BaseLayer, idx: number, a: BaseLayer[]) => void): void
+    getGroupsAndLayers(lyr: PluggableMap, filterFn?: (l: BaseLayer, idx: number, a: BaseLayer[]) => boolean): any[];
```

Comment:
Can `getGroupsAndLayers` return https://openlayers.org/en/latest/apidoc/module-ol_layer_Base-BaseLayer.html which is the common ancestor for layers and groups?

Reply:
I'm not sure that would work right.  For example, if we had a [ol/layer/Vector](https://openlayers.org/en/latest/apidoc/module-ol_layer_Vector-VectorLayer.html) that was returned, we could expect to call `getSource()`. Since this method doesn't exist on BaseLayer, typescript would get upset.

This one is difficult because there are so many classes that extend BaseLayer. I don't particularly like using `any`.  A more complete return type would be the following however it also has its own drawbacks.

`(BaseImage | BaseTile | BaseVector | Graticule | Group | HeatMap | Image | Layer | Tile | Vector)[]`

```
const myGroupsAndLayers = foo.getGroupsAndLayers(myMap, myFilter);
myGroupsAndLayers.forEach((element) => {
  // typescript would be helpful here is knowing the types that all elements share i.e. BaseLayer methods
  // We would still be unable to call getSource() here because we may not have a type that has this method
});
```

Reply:
I pushed a change for this.  let me know what you think.